CppImport/src/macro/MacroImportHelper.h

```diff
@@ -0,0 +1,120 @@
+ **      following disclaimer in the documentation and/or other materials provided with the distribution.
+ **    * Neither the name of the ETH Zurich nor the names of its contributors may be used to endorse or promote products
+ **      derived from this software without specific prior written permission.
+ **
+ **
+ ** THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES,
+ ** INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ ** DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ ** SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ ** SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+ ** WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ ** OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ **
+ **********************************************************************************************************************/
+
+#pragma once
+
+#include "cppimport_api.h"
+
+#include "ClangHelper.h"
+#include "AstMapping.h"
+#include "MacroExpansion.h"
+#include "NodeMapping.h"
+#include "MacroArgumentLocation.h"
+#include "MacroArgumentInfo.h"
+#include "ExpansionManager.h"
+#include "DefinitionManager.h"
+#include "LexicalHelper.h"
+#include "XMacroManager.h"
+#include "MetaDefinitionManager.h"
+#include "OOModel/src/allOOModelNodes.h"
+#include "clang/Lex/MacroArgs.h"
+
+namespace CppImport {
+
+class CPPIMPORT_API MacroImportHelper
+{
+	public:
+		MacroImportHelper(OOModel::Project* project);
+
+		/**
+		 * invoked after every imported translation unit to perform macro import.
+		 */
+		void macroGeneration();
+
+		/**
+		 * insert top level meta calls and remove nodes generated from top level expansions.
+		 * invoked after all translation units have been processed.
+		 */
+		void finalize();
+
+		void setSourceManager(const clang::SourceManager* sourceManager);
+		void setPreprocessor(const clang::Preprocessor* preprocessor);
+
+		void mapAst(clang::Stmt* clangAstNode, Model::Node* envisionAstNode);
+		void mapAst(clang::Decl* clangAstNode, Model::Node* envisionAstNode);
+
+		void addMacroDefinition(QString name, const clang::MacroDirective* md);
+		void addMacroExpansion(clang::SourceRange sr, const clang::MacroDirective* md, const clang::MacroArgs* args);
+
+	private:
+		OOModel::Project* root_;
+
+		ClangHelper clang_;
+		AstMapping astMapping_;
+		DefinitionManager definitionManager_;
+		ExpansionManager expansionManager_;
+		LexicalHelper lexicalHelper_;
+		XMacroManager xMacroManager_;
+		MetaDefinitionManager metaDefinitionManager_;
+		QHash<QString, OOModel::MetaCallExpression*> metaCalls_;
```

Comment:
I'm thinking, couldn't you move some of the functionality of the bigger classes (like this one) to some of those tiny structures that you had above? Of course only do this if it makes sense, as is the case with `MacroExpansion` and `topLevelNodes`.